### PR TITLE
Initial coordinator implementation

### PR DESCRIPTION
**Pull request by jwilder**
This PR adds a `Coordinator` type that handles writes to multiple `ShardWriter`s while ensuring a consistency level for each write request.  The `Coordinator` ensures that all the required shard groups and shards have been created before a write is executed.  When a write it executed, it attempts to write to each shard in a parallel.  If the there are multiple owners for shard, then each owner is written to in parallel as well.  The write will return as soon as the first error occurs or the write level consistency is achieved.

### CODE REVIEW

**Comment by pauldix** [comment]
You said:

> The write will return as soon as the first error occurs or the write level consistency is achieved.

Can you expand on this a bit? I would expect the write to return when the consistency level is achieved or we've timed out, no?

A single failure should be fine if they're asking for `any`, `one`, or `quorum` write level.

File: coordinator.go [hunk 1]
```diff
@@ -2,69 +2,252 @@ package influxdb
...
 			// return timeout error to caller
 			return ErrTimeout
 		case res := <-ch:
-			if !pol.IsDone(res.writerID, res.err) {
-				continue
-			}
-			if res.err != nil {
+			wrote += res.wrote
+
+			// ErrShardNotLocal might be returned from a local writer. Ignore it.
```
Comment (pauldix) [inline coordinator.go]: Why are local writers created for shards that aren't local? Seems odd to be doing this and returning an error that we're always ignoring.

File: data/points.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+// Point defines the values that will be written to the database
+type Point struct {
+	Name   string
+	Tags   Tags
+	Time   time.Time
+	Fields map[string]interface{}
+}
+
+func (p *Point) SeriesID() uint64 {
```
Comment (pauldix) [inline data/points.go]: I would rename this to something like `HashID`. `SeriesID` implies that it's something that is unique, which isn't guaranteed in this case.

**Comment by jwilder** [comment]
@pauldix I meant that `Write` call would return an error as soon as one of the `writeToShard` calls returns an error.  The `writeToShard` calls should only return an error when the requested consistency level could not be met.  It there are errors, it will wait for all of them to error our (or timeout) if the required responses has not been met before returning an error.

File: coordinator.go (hunk 1 above)
Comment (jwilder) [inline coordinator.go]: Agreed... will remove that.

File: data/points.go (hunk 2 above)
Comment (jwilder) [inline data/points.go]: Will fix.

**Comment by pauldix** [comment]
@jwilder cool, makes sense. If one of the `writeToShard` calls fails to hit the required consistency level and there are multiple shards being written to in the request, it makes sense that the entire write should fail out then.

File: coordinator.go [hunk 3]
```diff
@@ -2,69 +2,252 @@ package influxdb
...
+			return nil, err
+		}
+		timeRanges[t] = g
+	}
+
+	shardMapping := NewShardMapping()
+	for _, p := range wp.Points {
+		g := timeRanges[p.Time.Truncate(rp.ShardGroupDuration)]
+		sid := p.SeriesID()
+		shardInfo := g.Shards[sid%uint64(len(g.Shards))]
```
Comment (pauldix) [inline coordinator.go]: Should we extract this logic to a method `ShardGroupInfo` struct? We'll need to use it else where I think.

**Comment by pauldix** [comment]
What do you think of dropping the changes to `shard.go` and instead defining interfaces on the client side and mocking the shard out? This is only for the local writer. I was thinking of something like this:

```go
type PointsWriter interface {
  WritePoints([]Point) error
}
```

The reading back of the data is something that the tests for `shard` should do. Your writer test shouldn't be querying out, only mocking out that something can take a write and succeed or return an error so you can test the error path for LocalWriter.

**Comment by jwilder** [comment]
@pauldix The coordinator currently uses this interfaces for writes:

```
// ShardWriter provides the ability to write a slice of points to s given shard ID.
// It should return the number of times the set of points was written or an error
// if the write failed.
type ShardWriter interface {
	WriteShard(shardID uint64, points []data.Point) (int, error)
}
```

I was assuming that the `data.Node` and `ClusterWriter` would implement this interface.  `data.Node` would manage the local shards and writes to them. `ClusterWriter` would manage the remote connections to other data nodes and the writes to them as well.

I need to be able to specify the shard to write to and get some response back as to the number times the writes to that shard succeeded in order to ensure the consistency level.  I don't have any issue if a `Shard` implemented a `PointsWriter` interfaced behind the scenes though as I'm assuming I'd have a `data.Node` via a `ShardWriter` right now.

File: coordinator.go (hunk 3 above)
Comment (jwilder) [inline coordinator.go]: Sounds good.  Will fix.

**Comment by jwilder** [comment]
I can also remove the reading from the shard change.  That was stubbed out for tests initially but I don't think it's actually used now.  I'll remove it since it uses a mock `ShardWriter` to verify writes in the tests now.

**Comment by pauldix** [comment]
@jwilder that makes sense. I guess I was looking at the `shard_test` and `shard`. For those actual implementations, I'll do the shard stuff. You should be able to do a full test on this stuff without touching shard if you define that `PointsWriter` interface for what the data node uses to write to the actual shard.

**Pull request merged**